File: packages/typescript-core/src/committee.ts
// Threshold-decrypt orchestration: networking and quorum only; crypto is in the wasm core.

import { Aad, aadBytes } from "./aad.js";
import { cryptoProtocolVersion, lagrangeFor, maxCommitteeResponseBytes, openSecret } from "./crypto.js";
import type { Signer } from "./signer.js";
import type { PartialInput } from "./types.js";
import { fromHex, secretIdToHex, toHex } from "./util.js";

/** A committee node's `/health` response. */
export interface Health {
  status: string;
  epoch?: number;
  crypto_protocol_version?: number;
}

/** `/partial-decrypt` request body (matches the on-chain wire type). */
export interface PartialDecryptRequest {
  secret_id: string;
  subset: number[];
  lagrange_coeff: string;
  requester: string;
  block_hash: string;
  signature: string;
  auth: "substrate" | "ethereum";
  eth_address?: string;
  valid_until?: number;
  eth_signature?: string;
}

/** `/partial-decrypt` response body. */
export interface PartialDecryptResponse {
  node_index: number;
  partial: string;
  proof: string;
  crypto_protocol_version?: number;
  served_epoch?: number;
  shared_a?: string | null;
  joint_pk?: string | null;
  served_threshold?: number;
}

/** How the SDK reaches committee nodes. Swap in a fake for tests. */
export interface Transport {
  health(endpoint: string): Promise<Health>;
  partialDecrypt(endpoint: string, req: PartialDecryptRequest): Promise<PartialDecryptResponse>;
}

/** Bounds on each committee request, so one node can neither stall nor flood a decrypt. */
export interface FetchTransportOptions {
  /** Whole-request deadline in milliseconds. Default 30 000. */
  timeoutMs?: number;
  /** Cap on a response body in bytes. Default: the core's shared cap. */
  maxResponseBytes?: number;
}

/** A `fetch`-based transport for browsers and Node 22+. */
export class FetchTransport implements Transport {
  private readonly timeoutMs: number;
  private readonly maxResponseBytes: number;

  constructor(
    private readonly fetchImpl: typeof fetch = fetch,
    options: FetchTransportOptions = {},
  ) {
    this.timeoutMs = options.timeoutMs ?? 30_000;
    this.maxResponseBytes = options.maxResponseBytes ?? maxCommitteeResponseBytes();
  }

  async health(endpoint: string): Promise<Health> {
    const res = await this.fetchImpl(`${trimEnd(endpoint)}/health`, {
      signal: AbortSignal.timeout(this.timeoutMs),
    });
    if (!res.ok) throw new DecryptError("transport", `health ${res.status} from ${endpoint}`);
    return this.readJson<Health>(res, endpoint);
  }

  async partialDecrypt(endpoint: string, req: PartialDecryptRequest): Promise<PartialDecryptResponse> {
    const res = await this.fetchImpl(`${trimEnd(endpoint)}/partial-decrypt`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(req),
      signal: AbortSignal.timeout(this.timeoutMs),
    });
    if (!res.ok) throw new DecryptError("transport", `partial-decrypt ${res.status} from ${endpoint}`);
    return this.readJson<PartialDecryptResponse>(res, endpoint);
  }

  /** JSON-decode a body, refusing one over the cap instead of buffering it. */
  private async readJson<T>(res: Response, endpoint: string): Promise<T> {
    const tooLarge = () =>
      new DecryptError("transport", `response from ${endpoint} exceeded ${this.maxResponseBytes} bytes`);
    const declared = Number(res.headers.get("content-length"));
    if (declared > this.maxResponseBytes) throw tooLarge();
    const chunks: Uint8Array[] = [];
    let total = 0;
    const reader = res.body?.getReader();
    for (;;) {
      const next = reader ? await reader.read() : { done: true as const, value: undefined };
      if (next.done) break;
      total += next.value.length;
      if (total > this.maxResponseBytes) {
        await reader!.cancel();
        throw tooLarge();
      }
      chunks.push(next.value);
    }
    const body = new Uint8Array(total);
    chunks.reduce((offset, chunk) => (body.set(chunk, offset), offset + chunk.length), 0);
    return JSON.parse(new TextDecoder().decode(body)) as T;
  }
}

/** One committee node, with the chain-derived data a decryptor needs. */
export interface CommitteeNode {
  /** 1-based DKG evaluation point. */
  index: number;
  /** Base URL, no trailing `/partial-decrypt`. */
  endpoint: string;
  /** Bincode `FeldmanCommitment` for the secret's epoch, read from chain. */
  shareCommitment: Uint8Array;
}

/** Everything needed to recover one secret (chain-derived fields supplied by you). */
export interface DecryptParams {
  secretId: bigint;
  epoch: number;
  bindingId: Uint8Array;
  aad: Aad | Uint8Array;
  capsule: Uint8Array;
  ct: Uint8Array;
  sharedA: Uint8Array;
  blockHash: Uint8Array;
  threshold: number;
  nodes: CommitteeNode[];
}

/** Distinct failure causes, so callers branch on `.kind` instead of message text. */
export type DecryptErrorKind = "quorum" | "epoch" | "transport" | "crypto";

/** Where in the decrypt round trip a node stopped being usable. */
export type FaultStage =
  | "health"
  | "inactive"
  | "partial-decrypt"
  | "epoch-mismatch"
  | "protocol-version";

/**
 * Absent or `0` (node predates the field) is accepted: the version tag on every
 * proof still binds the transcript.
 */
function speaksOurProtocol(reported: number | undefined): boolean {
  return !reported || reported === cryptoProtocolVersion();
}

/** A uniform random integer in `[0, bound)` from the platform CSPRNG. */
function randomBelow(bound: number): number {
  // Rejection sampling avoids modulo bias.
  const limit = Math.floor(0x1_0000_0000 / bound) * bound;
  const word = new Uint32Array(1);
  for (;;) {
    crypto.getRandomValues(word);
    if (word[0]! < limit) return word[0]! % bound;
  }
}

/**
 * Choose `threshold` of `available` uniformly at random, returned in index order
 * (the signed subset's order).
 *
 * Random so no single node sees, or can deny, every decrypt. `random` is
 * injectable for tests only.
 */
export function chooseQuorum(
  available: readonly CommitteeNode[],
  threshold: number,
  random: (bound: number) => number = randomBelow,
): CommitteeNode[] {
  const pool = [...available];
  // Partial Fisher–Yates: the first `threshold` slots end up a uniform sample.
  for (let i = 0; i < threshold; i++) {
    const j = i + random(pool.length - i);
    [pool[i], pool[j]] = [pool[j]!, pool[i]!];
  }
  return pool.slice(0, threshold).sort((a, b) => a.index - b.index);
}

/** Why one committee node did not contribute to a quorum. Never carries request material. */
export interface NodeFault {
  index: number;
  endpoint: string;
  stage: FaultStage;
  /** The underlying reason, already rendered. */
  detail: string;
}

/** Appended to the message so a caller that only logs `err.message` still sees faults. */
function summarizeFaults(faults: readonly NodeFault[]): string {
  if (faults.length === 0) return "";
  return (
    " — " +
    faults.map((f) => `node ${f.index} (${f.endpoint}) ${f.stage}: ${f.detail}`).join("; ")
  );
}

function reasonOf(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

/** A typed decrypt failure. */
export class DecryptError extends Error {
  constructor(
    readonly kind: DecryptErrorKind,
    message: string,
    /** For `kind === "quorum"`, one entry per dropped node. */
    readonly faults: readonly NodeFault[] = [],
  ) {
    super(message);
    this.name = "DecryptError";
  }
}

/**
 * Recover a secret from a threshold quorum: health-probe, pick `threshold` active
 * nodes at random, sign per node, query each, then verify + aggregate + AEAD-open.
 *
 * A failing node is dropped and the quorum re-formed; `threshold` nodes agreeing
 * on another served epoch throws `kind: "epoch"` (refetch and retry).
 *
 * Returns the recovered plaintext: never log it, and {@link wipe} it when done.
 */
export async function decrypt(
  transport: Transport,
  signer: Signer,
  params: DecryptParams,
): Promise<Uint8Array> {
  const faults: NodeFault[] = [];
  const probes = await Promise.allSettled(
    params.nodes.map(async (node) => ({ node, health: await transport.health(node.endpoint) })),
  );
  const active: CommitteeNode[] = [];
  probes.forEach((probe, i) => {
    const node = params.nodes[i]!;
    if (probe.status === "rejected") {
      faults.push({
        index: node.index,
        endpoint: node.endpoint,
        stage: "health",
        detail: reasonOf(probe.reason),
      });
    } else if (!speaksOurProtocol(probe.value.health.crypto_protocol_version)) {
      faults.push({
        index: node.index,
        endpoint: node.endpoint,
        stage: "protocol-version",
        detail: `speaks crypto protocol v${probe.value.health.crypto_protocol_version}, this SDK speaks v${cryptoProtocolVersion()}`,
      });
    } else if (probe.value.health.status !== "active") {
      const h = probe.value.health;
      faults.push({
        index: node.index,
        endpoint: node.endpoint,
        stage: "inactive",
        detail: `status ${JSON.stringify(h.status)}, epoch ${h.epoch ?? 0}, crypto protocol v${h.crypto_protocol_version ?? 0}`,
      });
    } else {
      active.push(node);
    }
  });

  if (active.length < params.threshold) {
    throw new DecryptError(
      "quorum",
      `quorum unavailable: need ${params.threshold} healthy nodes, found ${active.length}${summarizeFaults(faults)}`,
      faults,
    );
  }

  // Each fault removes a node, so the loop terminates.
  let available = active;
  const rotatedVotes = new Map<number, number>();

  while (available.length >= params.threshold) {
    const chosen = chooseQuorum(available, params.threshold);
    const subset = chosen.map((n) => n.index);

    // Sign per node: the payload binds the recipient, so it can't replay to a peer.
    const partials: PartialInput[] = [];
    let faulty: number | null = null;

    for (const node of chosen) {
      const lambda = lagrangeFor(node.index, subset);
      const auth = await signer.authorize({
        secretId: params.secretId,
        subset,
        recipientIndex: node.index,
        blockHash: params.blockHash,
      });
      const req: PartialDecryptRequest = {
        secret_id: secretIdToHex(params.secretId),
        subset,
        lagrange_coeff: toHex(lambda),
        requester: auth.requester,
        block_hash: toHex(params.blockHash),
        signature: auth.signature,
        auth: auth.auth,
        eth_address: auth.eth_address,
        valid_until: auth.valid_until,
        eth_signature: auth.eth_signature,
      };

      let resp: PartialDecryptResponse;
      try {
        resp = await transport.partialDecrypt(node.endpoint, req);
      } catch (e) {
        faults.push({
          index: node.index,
          endpoint: node.endpoint,
          stage: "partial-decrypt",
          detail: reasonOf(e),
        });
        faulty = node.index;
        break;
      }

      // `threshold` votes for the same new epoch is a real rotation; fewer is one
      // misbehaving node. 0/undefined means an older node serving the current epoch.
      if (resp.served_epoch && resp.served_epoch !== params.epoch) {
        const votes = (rotatedVotes.get(resp.served_epoch) ?? 0) + 1;
        rotatedVotes.set(resp.served_epoch, votes);
        if (votes >= params.threshold) {
          throw new DecryptError(
            "epoch",
            `secret served under epoch ${resp.served_epoch}, state supplied for ${params.epoch}; refetch and retry`,
          );
        }
        faults.push({
          index: node.index,
          endpoint: node.endpoint,
          stage: "epoch-mismatch",
          detail: `served epoch ${resp.served_epoch}, state supplied for ${params.epoch}`,
        });
        faulty = node.index;
        break;
      }

      partials.push({
        point: node.index,
        partial: fromHex(resp.partial),
        proof: fromHex(resp.proof),
        commitment: node.shareCommitment,
      });
    }

    if (faulty !== null) {
      available = available.filter((n) => n.index !== faulty);
      continue;
    }
    return openOrThrow(params, partials);
  }

  // Out of nodes: report the most-voted divergent epoch as a rotation, if any.
  let dominant: { epoch: number; votes: number } | null = null;
  for (const [epoch, votes] of rotatedVotes) {
    if (!dominant || votes > dominant.votes) dominant = { epoch, votes };
  }
  if (dominant) {
    throw new DecryptError(
      "epoch",
      `secret served under epoch ${dominant.epoch}, state supplied for ${params.epoch}; refetch and retry`,
    );
  }
  throw new DecryptError(
    "quorum",
    `quorum unavailable: need ${params.threshold} healthy nodes, found ${available.length}${summarizeFaults(faults)}`,
    faults,
  );
}

function openOrThrow(params: DecryptParams, partials: PartialInput[]): Uint8Array {
  try {
    return openSecret({
      sharedA: params.sharedA,
      capsule: params.capsule,
      secretId: params.secretId,
      epoch: params.epoch,
      bindingId: params.bindingId,
      aad: aadBytes(params.aad),
      ct: params.ct,
      partials,
    });
  } catch (e) {
    throw new DecryptError("crypto", e instanceof Error ? e.message : String(e));
  }
}
function trimEnd(url: string): string {
  return url.replace(/\/+$/, "");
}

File: crates/matter-sdk/src/committee.rs
//! Threshold-decrypt orchestration: choose a quorum, sign per node, collect
//! partials. All crypto stays in [`matter_sdk_core`].

use matter_sdk_core::wire::{from_0x, secret_id_to_hex, to_0x, PartialDecryptRequest};
use matter_sdk_core::{
    lagrange_for,
    open_secret,
    PartialInput,
    Plaintext,
    CRYPTO_PROTOCOL_VERSION,
};
use rand::rngs::OsRng;
use rand::seq::SliceRandom;
use rand::Rng;

use crate::error::{FaultStage, NodeFault, Result, SdkError};
use crate::signer::{Signer, SigningRequest};
use crate::transport::Transport;

/// One committee node the SDK can query, with the on-chain data a decryptor needs.
#[derive(Debug, Clone)]
pub struct CommitteeNode {
    /// The node's 1-based DKG evaluation point (`dkg_index`).
    pub index: u64,
    /// The node's public base URL (no trailing `/partial-decrypt`).
    pub endpoint: String,
    /// Bincode `FeldmanCommitment` (`g_j`) for the secret's epoch.
    pub share_commitment: Vec<u8>,
}

/// Everything needed to recover one secret. The caller reads `shared_a`,
/// `threshold`, node endpoints and per-node `share_commitment` from chain with
/// its own Substrate client.
#[derive(Debug, Clone)]
pub struct DecryptRequest<'a> {
    /// The secret id.
    pub secret_id: u128,
    /// The epoch the secret was sealed under; `shared_a` and commitments must match.
    pub epoch: u32,
    /// The binding id from the on-chain envelope.
    pub binding_id: &'a [u8],
    /// The AAD the secret was sealed under (use a [`matter_sdk_core::Aad`] tag).
    pub aad: &'a [u8],
    /// Bincode capsule from the envelope.
    pub capsule: &'a [u8],
    /// The `nonce ‖ ciphertext` from the envelope.
    pub ct: &'a [u8],
    /// Bincode `shared_a` for `epoch`.
    pub shared_a: &'a [u8],
    /// A recent finalized block hash; the freshness anchor for request signatures.
    pub block_hash: [u8; 32],
    /// The committee threshold `t` for `epoch`.
    pub threshold: usize,
    /// Candidate nodes; at least `threshold` must be healthy.
    pub nodes: &'a [CommitteeNode],
}

/// `0` (field absent) is accepted: the version tag on every proof still binds
/// the transcript.
fn speaks_our_protocol(reported: u16) -> bool {
    reported == 0 || reported == CRYPTO_PROTOCOL_VERSION
}

/// Choose `threshold` of `available` uniformly at random, in index order.
///
/// Random so no single node sees, or can deny, every decrypt.
fn choose_quorum<'n, R: Rng + ?Sized>(
    available: &[&'n CommitteeNode],
    threshold: usize,
    rng: &mut R,
) -> Vec<&'n CommitteeNode> {
    let mut chosen: Vec<&CommitteeNode> =
        available.choose_multiple(rng, threshold).copied().collect();
    chosen.sort_by_key(|n| n.index);
    chosen
}

/// Recover a secret from a threshold quorum of partial decryptions.
///
/// Health-probes nodes, picks `threshold` active ones at random, signs per node,
/// then verifies, aggregates, and AEAD-opens. A failing node is dropped and the
/// quorum re-formed. The returned [`Plaintext`] is a zeroizing buffer.
///
/// Errors: [`SdkError::QuorumUnavailable`] (with a [`NodeFault`] per dropped
/// node) if too few nodes are usable; [`SdkError::EpochRotated`] if the
/// committee served a different epoch (refetch and retry); [`SdkError::Core`]
/// with [`matter_sdk_core::CoreError::Aead`] if the secret cannot be opened.
pub async fn decrypt<T, S>(transport: &T, signer: &S, req: &DecryptRequest<'_>) -> Result<Plaintext>
where
    T: Transport,
    S: Signer,
{
    // 1. Health-probe sequentially (committees are small); record why each
    //    dropped node dropped.
    let mut active: Vec<&CommitteeNode> = Vec::new();
    let mut faults: Vec<NodeFault> = Vec::new();
    for node in req.nodes {
        match transport.health(&node.endpoint).await {
            Ok(health) if !speaks_our_protocol(health.crypto_protocol_version) => {
                faults.push(NodeFault {
                    index: node.index,
                    endpoint: node.endpoint.clone(),
                    stage: FaultStage::ProtocolVersion,
                    detail: format!(
                        "speaks crypto protocol v{}, this SDK speaks v{CRYPTO_PROTOCOL_VERSION}",
                        health.crypto_protocol_version
                    ),
                })
            }
            Ok(health) if health.is_active() => active.push(node),
            Ok(health) => faults.push(NodeFault {
                index: node.index,
                endpoint: node.endpoint.clone(),
                stage: FaultStage::Inactive,
                detail: format!(
                    "status {:?}, epoch {}, crypto protocol v{}",
                    health.status, health.epoch, health.crypto_protocol_version
                ),
            }),
            Err(e) => faults.push(NodeFault {
                index: node.index,
                endpoint: node.endpoint.clone(),
                stage: FaultStage::Health,
                detail: e.to_string(),
            }),
        }
    }
    if active.len() < req.threshold {
        return Err(SdkError::QuorumUnavailable {
            needed: req.threshold,
            active: active.len(),
            faults,
        });
    }

    // 2. A node that fails or serves a different epoch is dropped and the subset
    //    re-formed, so one node cannot deny the decrypt. A genuine rotation is
    //    `threshold` nodes agreeing on the same new epoch. Each fault shrinks
    //    `available`, so the loop terminates.
    let mut available = active;
    let mut rotated_votes: std::collections::BTreeMap<u32, usize> =
        std::collections::BTreeMap::new();

    while available.len() >= req.threshold {
        let chosen = choose_quorum(&available, req.threshold, &mut OsRng);
        let subset: Vec<u64> = chosen.iter().map(|n| n.index).collect();

        let mut partials: Vec<PartialInput> = Vec::with_capacity(chosen.len());
        let mut faulty: Option<u64> = None;
        for node in &chosen {
            let lambda = lagrange_for(node.index, &subset)?;

            // The payload binds this node's index, so it cannot replay the auth
            // to a peer. The signer never reveals its key.
            let auth = signer.authorize(&SigningRequest {
                secret_id: req.secret_id,
                subset: &subset,
                recipient_index: node.index,
                block_hash: req.block_hash,
                valid_until: None,
            })?;

            let request = PartialDecryptRequest {
                secret_id: secret_id_to_hex(req.secret_id),
                subset: subset.clone(),
                lagrange_coeff: to_0x(&lambda),
                requester: auth.requester,
                block_hash: to_0x(&req.block_hash),
                signature: auth.signature,
                auth: auth.auth,
                eth_address: auth.eth_address,
                valid_until: auth.valid_until,
                eth_signature: auth.eth_signature,
            };

            let resp = match transport.partial_decrypt(&node.endpoint, &request).await {
                Ok(resp) => resp,
                Err(e) => {
                    faults.push(NodeFault {
                        index: node.index,
                        endpoint: node.endpoint.clone(),
                        stage: FaultStage::PartialDecrypt,
                        detail: e.to_string(),
                    });
                    faulty = Some(node.index);
                    break;
                }
            };

            // `threshold` votes for the same new epoch is a rotation; fewer is one
            // misbehaving node. 0 means an older node serving the current epoch.
            if resp.served_epoch != 0 && resp.served_epoch != req.epoch {
                let votes = rotated_votes.entry(resp.served_epoch).or_insert(0);
                *votes += 1;
                if *votes >= req.threshold {
                    return Err(SdkError::EpochRotated {
                        served: resp.served_epoch,
                        provided: req.epoch,
                    });
                }
                faults.push(NodeFault {
                    index: node.index,
                    endpoint: node.endpoint.clone(),
                    stage: FaultStage::EpochMismatch,
                    detail: format!(
                        "served epoch {}, state supplied for {}",
                        resp.served_epoch, req.epoch
                    ),
                });
                faulty = Some(node.index);
                break;
            }

            partials.push(PartialInput {
                point: node.index,
                partial: from_0x("partial", &resp.partial)?,
                proof: from_0x("proof", &resp.proof)?,
                commitment: node.share_commitment.clone(),
            });
        }

        match faulty {
            Some(bad) => {
                available.retain(|n| n.index != bad);
                continue;
            }
            // 3. Verify, aggregate, and AEAD-open in the pure core.
            None => {
                return Ok(open_secret(
                    req.shared_a,
                    req.capsule,
                    req.secret_id,
                    req.epoch,
                    req.binding_id,
                    req.aad,
                    req.ct,
                    &partials,
                )?);
            }
        }
    }

    // Out of good nodes: report the most-voted divergent epoch as a rotation.
    if let Some((&served, _)) = rotated_votes.iter().max_by_key(|(_, &votes)| votes) {
        return Err(SdkError::EpochRotated {
            served,
            provided: req.epoch,
        });
    }
    Err(SdkError::QuorumUnavailable {
        needed: req.threshold,
        active: available.len(),
        faults,
    })
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeSet;

    use rand::rngs::StdRng;
    use rand::SeedableRng;

    use super::*;

    fn nodes(n: u64) -> Vec<CommitteeNode> {
        (1..=n)
            .map(|index| CommitteeNode {
                index,
                endpoint: format!("http://node-{index}"),
                share_commitment: Vec::new(),
            })
            .collect()
    }

    #[test]
    fn a_quorum_is_t_distinct_available_nodes_in_index_order() {
        let all = nodes(7);
        let available: Vec<&CommitteeNode> = all.iter().collect();
        let mut rng = StdRng::seed_from_u64(7);
        for _ in 0..100 {
            let chosen = choose_quorum(&available, 4, &mut rng);
            let indices: Vec<u64> = chosen.iter().map(|n| n.index).collect();
            assert_eq!(indices.len(), 4);
            assert!(
                indices.windows(2).all(|w| w[0] < w[1]),
                "sorted and distinct: {indices:?}"
            );
        }
    }

    #[test]
    fn quorum_selection_is_not_the_fixed_lowest_indices() {
        let all = nodes(5);
        let available: Vec<&CommitteeNode> = all.iter().collect();
        let mut rng = StdRng::seed_from_u64(42);
        let draws: Vec<Vec<u64>> = (0..200)
            .map(|_| {
                choose_quorum(&available, 3, &mut rng)
                    .iter()
                    .map(|n| n.index)
                    .collect()
            })
            .collect();
        let distinct: BTreeSet<&Vec<u64>> = draws.iter().collect();
        assert!(distinct.len() > 1, "every draw was {:?}", draws[0]);
        for index in 1..=5u64 {
            let hits = draws.iter().filter(|d| d.contains(&index)).count();
            // Expected 120/200; 60 is a loose floor a skewed choice cannot meet.
            assert!(hits >= 60, "node {index} chosen only {hits}/200 times");
        }
    }
}
